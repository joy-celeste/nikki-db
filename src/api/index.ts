import { HttpRequestException } from '../models/Errors';
import database from '../Database';

export const fetchItemData = async (itemId: number): Promise<any> => {
  try {
    return database.collection('clothes').doc(itemId.toString())
      .get()
      .then((doc) => {
        if (!doc.exists) {
          console.log('No such document!');
          return null;
        }
        console.log('Retrieved data from database.', doc.data());
        return doc.data();
      });
  } catch (error) {
    throw new HttpRequestException(error.code, 'Error getting document.');
  }
};

export default fetchItemData;
