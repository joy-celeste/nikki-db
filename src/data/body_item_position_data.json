{"position":{"30":{"pot_scale":1,"posy":62,"posx":-7.5},"24":{"pot_scale":1,"posy":62,"posx":-9.5},"23":{"pot_scale":1,"posy":257.5,"posx":-18},"22":{"pot_scale":1,"posy":62,"posx":-7.5},"17":{"pot_scale":1,"posy":254,"posx":-23},"25":{"pot_scale":1,"posy":254,"posx":-23},"6":{"posy":299.5,"posx":-23.5},"7":{"posy":376,"posx":-30.5},"8":{"posy":260,"posx":-24},"9":{"posy":146,"posx":-22.5},"29":{"pot_scale":1,"posy":258,"posx":-12.5},"28":{"pot_scale":1,"posy":66.5,"posx":-11},"27":{"pot_scale":1,"posy":238.5,"posx":-26},"26":{"pot_scale":1,"posy":73.5,"posx":-7},"16":{"pot_scale":1,"posy":62,"posx":-9.5},"19":{"pot_scale":1,"posy":238.5,"posx":-26},"18":{"pot_scale":1,"posy":73.5,"posx":-7},"1":{"posy":-29.5,"posx":-22.5},"2":{"posy":239.5,"posx":-11.5},"3":{"posy":128.5,"posx":-21},"4":{"posy":253.5,"posx":-11.5},"5":{"posy":56,"posx":-9.5},"13":{"posy":253.5,"posx":-11.5},"12":{"posy":205.5,"posx":-24},"15":{"pot_scale":1,"posy":257.5,"posx":-18},"14":{"posy":56,"posx":-9.5},"10":{"posy":-204.5,"posx":-9.5},"11":{"posy":417.5,"posx":-21.5},"21":{"pot_scale":1,"posy":258,"posx":-12.5},"20":{"pot_scale":1,"posy":66.5,"posx":-11}},"id":1}