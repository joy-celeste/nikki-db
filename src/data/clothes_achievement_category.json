{"8":"Troupe","9":"4 Seasons","5":"Others","4":"Styling Contest","7":"Festivals","6":"7 Nations","3":"Journey","21":"Museum","20":"御苑琼芳","23":"Constellation Hall","22":"Ancient Fossil Hall","15":"North","14":"Pigeon","17":"Ruin","16":"Wasteland","18":"Happiness","19":"Story Suit","2":"Collection","1":"My Growth","12":"Lilith","13":"Cloud","10":"Stars","11":"Apple"}