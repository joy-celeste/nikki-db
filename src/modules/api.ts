import { HttpRequestException } from './errors';
import database from './database';
import { API_CONSTANTS } from './constants';

export const fetchItemData = async (itemId: number): Promise<any> =>
  database.collection(API_CONSTANTS.CLOTHES).doc(itemId.toString())
    .get()
    .then((response) => {
      if (!response) {
        console.log('No such document!');
        return null;
      }
      console.log('Retrieved data from database.');
      return response.data();
    })
    .catch((error) => {
      console.log(new HttpRequestException(error, 'Error getting response.'));
    });

export {};
