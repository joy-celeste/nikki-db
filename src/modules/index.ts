import { combineReducers } from 'redux';
import { userReducer } from './user';
import { productsReducer } from './products';

export const rootReducer = combineReducers({
  user: userReducer,
  products: productsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;