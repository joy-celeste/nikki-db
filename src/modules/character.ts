import { Dispatch } from 'react';
import { AnyAction } from 'redux';
import { RootState } from '.';
import { ACTION_CONSTANTS, DEFAULT_BODY, DEFAULT_CLOTHES, DEFAULT_AMPUTATIONS, SUBTYPES, BODY } from './constants';
import { AmputationData, AmputationParts, ItemData, ItemId, ItemsData, SubType } from './data';
import { Item } from './item';

export type BodyPart = number;
export type BodyParts = Set<BodyPart>;
export type Clothes = Record<SubType, ItemId>;
export type Amputations = Record<AmputationParts, ItemId[]>;

export class Character {
  visibleParts: BodyParts;
  clothes: Clothes;
  amputations: Amputations;

  constructor(input?: any) {
    this.visibleParts = input && input.visibleParts ? new Set(input.visibleParts) : new Set(DEFAULT_BODY);
    this.clothes = input && input.clothes ? { ...input.clothes } : { ...DEFAULT_CLOTHES };
    this.amputations = input && input.amputations ? { ...input.amputations } : { ...DEFAULT_AMPUTATIONS };
  }

  hide(bodyPart: BodyPart): void {
    if (this.visibleParts.has(bodyPart)) {
      this.visibleParts.delete(bodyPart);
    }
  }

  show(bodyPart: BodyPart): void {
    this.visibleParts.add(bodyPart);
  }

  wear(itemData: ItemData): void {
    const { subType, id, amputationData} = itemData;

    if (this.clothes[subType] && this.clothes[subType] === id && subType !== SUBTYPES.HAIR) {
      this.remove(subType);
      return;
    }

    if (this.clothes[subType] && this.clothes[subType] !== id) {
      this.remove(subType);
    }

    if (subType === SUBTYPES.DRESS) {
      this.remove(SUBTYPES.TOP);
      this.remove(SUBTYPES.BOTTOM);
    } else if (subType === SUBTYPES.TOP) {
      this.remove(SUBTYPES.DRESS);
    } else if (subType === SUBTYPES.BOTTOM) {
      this.remove(SUBTYPES.DRESS);
    }

    this.clothes[subType] = id;
    this.updateAmputations(id, amputationData, subType);
    this.updateVisibleBodyParts();
  }

  remove(subtype: SubType): void {
    if (this.clothes[subtype]) {
      const foundItem : ItemId = this.clothes[subtype];
      delete this.clothes[subtype];

      Object.keys(this.amputations).forEach((bodyPartStr) => {
        const bodyPart = parseInt(bodyPartStr, 10) as AmputationParts;
        this.amputations[bodyPart] = this.amputations[bodyPart].filter((itemId: ItemId) => itemId !== foundItem);
        if (this.amputations[bodyPart].length === 0) {
          this.show(bodyPart);
        }
      });
    }

    if (subtype === SUBTYPES.DRESS) {
      this.show(BODY.BRA);
      this.show(BODY.PANTY);
    } else if (subtype === SUBTYPES.TOP) {
      this.show(BODY.BRA);
    } else if (subtype === SUBTYPES.BOTTOM) {
      this.show(BODY.PANTY);
    }
  }

  updateAmputations(itemId: ItemId, amputationData: AmputationData, subtype: SubType): void {
    Object.keys(this.amputations).forEach((bodyPartStr) => {
      const bodyPart = parseInt(bodyPartStr, 10) as AmputationParts;
      if (this.shouldHideBodyPart(bodyPart, amputationData, subtype)) {
        this.amputations[bodyPart] = [...this.amputations[bodyPart], itemId];
      }
    });
  }

  shouldHideBodyPart(bodyPart: BodyPart, amputationData: AmputationData, subtype: SubType): boolean {
    if (amputationData && amputationData[bodyPart]) {
      return true;
    }

    if ((bodyPart === BODY.BRA || bodyPart === BODY.VEST) && (subtype === SUBTYPES.TOP || subtype === SUBTYPES.DRESS)) {
      return true;
    }

    if (bodyPart === BODY.PANTY && (subtype === SUBTYPES.BOTTOM || subtype === SUBTYPES.DRESS)) {
      return true;
    }

    return false;
  }

  // also refactor wear to only take in itemdata instead of having to use 3 params

  updateVisibleBodyParts(): void {
    // If any amputation data is associated with an amputation location, hide the limb
    // Else if nothing is being worn on this amputation location right now, make it visible
    Object.entries(this.amputations).forEach(([bodyPartStr, items]) => {
      const bodyPart = parseInt(bodyPartStr, 10) as BodyPart;
      if (items.length >= 1) {
        this.hide(bodyPart); // have to cast as int as Object.keys turns keys into string
      } else {
        this.show(bodyPart);
      }
    });
  }
}

export type CharacterState = {
  history: Character[];
  step: number;
};

const initialState: CharacterState = {
  history: [new Character()],
  step: 0,
};

// ACTIONS
const addToHistory = (character: Character, step: number): AnyAction => ({
  type: ACTION_CONSTANTS.CHARACTER_ADD_TO_HISTORY,
  payload: {
    character, step,
  },
});

// USE-CASE
export const wearItem = (itemId: ItemId) =>
  async(dispatch: Dispatch<AnyAction>, getState: () => RootState): Promise<any> => {
    const itemDataState: ItemsData = getState().data.itemsData;
    const itemData = itemDataState[itemId];

    if (itemData && itemData.subType) {
      const charState: CharacterState = getState().character;
      const oldChar: Character = charState.history[charState.step];
      const newChar: Character = new Character(oldChar);
      newChar.wear(itemData);
      dispatch(addToHistory(newChar, charState.step + 1));
    }
  };

// REDUCER
export function characterReducer(
  state = initialState,
  action: AnyAction,
): CharacterState {
  switch (action.type) {
    case ACTION_CONSTANTS.CHARACTER_ADD_TO_HISTORY:
      return {
        ...state,
        history: [...state.history, action.payload.character],
        step: action.payload.step,
      };
    default:
      return state;
  }
}
