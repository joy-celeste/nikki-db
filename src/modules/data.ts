import { AnyAction } from 'redux';
import { DocumentData } from '@firebase/firestore-types';
import { DeserializeNullException } from './errors';
import { ACTION_CONSTANTS, BODY, BODY_ITEM_DATA, BODY_ITEM_ID, BODY_PARTS_DEPTHS,
  DEFAULT_AMPUTATIONS_LIST, DEPTHTYPE_TO_SUBTYPES, SUBTYPES_LIST, CLOTHES_DATA } from './constants';
import { Character, CharacterState, wearItem, addToHistory } from './character';
import { RootState } from '.';
import { changeHiddenItemList } from './editor';

export type SubType = typeof SUBTYPES_LIST[number];
export type AmputationParts = typeof DEFAULT_AMPUTATIONS_LIST[number];
export type AmputationData = Record<AmputationParts, boolean>;
export type Depths = Record<number, number>;
export type ItemId = number;
export type DepthType = number;

export interface SubtypeInfo {
  subtype: number,
  depth: Record<number, number>
}

export class ItemData {
  id?: ItemId;
  name?: string;
  description?: string;
  depthType?: DepthType;
  subType?: SubType;
  amputationData?: AmputationData;
  position?: PositionData[];
  depths?: Depths;

  constructor(input?: DocumentData) {
    this.deserialize(input);
  }

  deserialize(input: DocumentData): ItemData {
    if (!input) {
      throw new DeserializeNullException('Cannot deserialize null input for ItemData');
    }

    this.id = input.id || null;
    this.name = input.name || '';
    this.description = input.desc || '';
    this.position = null;
    this.depthType = null;
    this.subType = null;
    this.depths = null;
    this.amputationData = null;

    if (input.amputation_data) {
      this.amputationData = {
        [BODY.ARM]: !!input.amputation_data.arm,
        [BODY.TORSO]: !!input.amputation_data.body,
        [BODY.BREAST]: !!input.amputation_data.body,
        [BODY.LEG]: !!input.amputation_data.leg,
        [BODY.PANTY]: !!input.amputation_data.panty,
      };
    }

    if (input.depth_type || input.id === BODY_ITEM_ID) {
      this.depthType = input.depth_type;
      const subtypeData = input.id !== BODY_ITEM_ID
        ? DEPTHTYPE_TO_SUBTYPES[input.depth_type.toString()]
        : BODY_PARTS_DEPTHS;
      this.subType = subtypeData.subtype;
      this.depths = subtypeData.depth;
    }

    if (input.position) {
      const position: PositionData[] = [];
      Object.entries(input.position).forEach(([key, value]) => {
        position.push(new PositionData(parseInt(key, 10), value, this.depths));
      });
      this.position = position;
    }
    return this;
  }
}

export class PositionData {
  depth: number;
  x: number;
  y: number;
  z: number;
  height?: number;
  width?: number;
  scale?: number;

  constructor(depth: number, input: DocumentData, depths?: Depths) {
    this.deserialize(depth, input, depths);
  }

  deserialize(depth: number, input: DocumentData, depths?: Depths): PositionData {
    if (!depth || !input) {
      throw new DeserializeNullException('Cannot deserialize null input for PositionData');
    }

    this.depth = depth;
    this.x = input.posx;
    this.y = input.posy;
    this.z = null;
    this.height = input.height || null;
    this.width = input.width || null;
    this.scale = input.pot_scale || null;

    if (depths) {
      this.calculateDepth(depth, depths);
    }
    return this;
  }

  // Don't ask me about this logic... I'm just copying this straight from the original devs
  calculateDepth(depth: number, depths: Depths): void {
    if (depth < 100) {
      this.z = depths[depth];
    } else {
      const key: number = depth / 100;
      const ceil: number = Math.ceil(key);
      /* istanbul ignore next */
      if (key <= 0 || ceil === key) { // These are for animated effects.
        this.z = depths[ceil] - 5;
      }
      this.z = depths[ceil - 1] - 5;
    }
  }
}

export const bodyItemData = new ItemData(BODY_ITEM_DATA);

export type ItemsData = Record<ItemId, ItemData>;
export type DataState = {
  itemsData: ItemsData;
  loading: boolean;
};

export const initialState: DataState = {
  itemsData: { 10001: new ItemData(CLOTHES_DATA[10001]) },
  loading: false,
};

// ACTIONS
export const addItemData = (itemId: number, itemData: ItemData): AnyAction => ({
  type: ACTION_CONSTANTS.DATA_ADD_ITEMS,
  payload: {
    itemId, itemData,
  },
});

export type BreakdownData = {
  variant: string;
  itemId: ItemId;
  prosthesis: boolean;
  subtype: SubType;
  name: string;
};

// USE-CASE
export const loadItem = (itemId: ItemId) =>
  async(dispatch: Function, getState: () => RootState): Promise<void> => {
    const items: ItemsData = getState().data.itemsData;
    if (!(itemId in items) && CLOTHES_DATA[itemId]) {
      dispatch(addItemData(itemId, new ItemData(CLOTHES_DATA[itemId])));
    } else if (itemId in items) {
      dispatch(wearItem(itemId));
    }
  };

export const loadMultipleItems = (itemIds: ItemId[]) =>
  async(dispatch: Function, getState: () => RootState): Promise<void> => {
    dispatch(changeHiddenItemList(new Set<ItemId>()));
    if (itemIds.length === 1) {
      dispatch(loadItem(itemIds[0]));
      return;
    }

    const items: ItemsData = getState().data.itemsData;
    const charState: CharacterState = getState().character;
    const newChar: Character = new Character();
    let itemData: ItemData;

    await Promise.allSettled(itemIds.map(async (itemId) => {
      if (!(itemId in items)) {
        itemData = new ItemData(CLOTHES_DATA[itemId]);
        dispatch(addItemData(itemId, itemData));
      } else {
        itemData = items[itemId];
      }
      newChar.wear(itemData);
    })).then(() => {
      dispatch(addToHistory(newChar, charState.step + 1));
    });
  };

// REDUCER
export function dataReducer(
  state = initialState,
  action: AnyAction,
): DataState {
  switch (action.type) {
    case ACTION_CONSTANTS.DATA_ADD_ITEMS:
      return {
        ...state,
        itemsData: {
          ...state.itemsData,
          [action.payload.itemId]: action.payload.itemData,
        },
      };
    default:
      return state;
  }
}
