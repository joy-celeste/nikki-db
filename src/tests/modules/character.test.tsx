import { Store } from 'redux';
import {
  posed_dress, posed_coat, posed_shoes, simple_dress, simple_hair, simple_hat, complex_hair, deer_spirit, motorcycle_spirit,
} from '../test_data/data';
import { DeserializeNullException } from '../../modules/errors';
import { RootState } from '../../modules';
import { AmputationData, AmputationParts, ItemData, ItemId, loadItem, PositionData, SubType } from '../../modules/data';
import { createStoreWithMiddleware } from '../helpers';
import * as api from '../../modules/api';
import { Amputations, Character, Clothes, Body } from '../../modules/character';
import { ACTION_CONSTANTS, DEFAULT_BODY, DEFAULT_CLOTHES, SUBTYPES } from '../../modules/constants';

const validateAmputations = (itemData: ItemData, character: Character) => {
  const amputations: Amputations = character.amputations;
  const clothes = new Set(Object.values(character.clothes));

  for (let strBodyPart of Object.keys(amputations)) {
    const bodyPart = parseInt(strBodyPart) as AmputationParts;

    // Validate that ITEMDATA'S amputation data was added if the amputation data exists
    // If it doesn't exist, validate that it doesn't show up in CHARACTER.AMPUTATIONS
    if (itemData.amputationData) {
      if (itemData.amputationData[bodyPart] && character.clothes[itemData.subType] === itemData.id) {
        expect(amputations[bodyPart].includes(itemData.id)).toBe(true);
      } else {
        expect(amputations[bodyPart].includes(itemData.id)).toBe(false);
      }
    } else {
      expect(amputations[bodyPart].includes(itemData.id)).toBe(false);
    }

    // Validate there is no extraneous/old/outdated data in CHARACTER.AMPUTATIONS
    amputations[bodyPart].forEach((itemId: ItemId) => {
      expect(clothes.has(itemId)).toBe(true);
    });
  }
}

describe('Character', () => {
  test('Initializes to default data if empty constructor', () => {
    const new_character: Character = new Character();
    expect(new_character.body).toEqual(DEFAULT_BODY);
    expect(new_character.clothes).toEqual(DEFAULT_CLOTHES);
    expect(new_character).toMatchSnapshot();
  });

  test('CHARACTER: wear() - success case: dress on new character', () => {
    const character: Character = new Character();
    const dress: ItemData = new ItemData(posed_dress);
    character.wear(dress.subType, dress.id, dress.amputationData);
    expect(Object.keys(character.clothes)).toHaveLength(2);
    expect(character.clothes.hasOwnProperty(dress.subType))
    expect(character.clothes[dress.subType]).toEqual(dress.id)
    expect(character).toMatchSnapshot();
  });

  test('CHARACTER: wear() - success: replacing default hair with something else', () => {
    const character: Character = new Character();
    const hair: ItemData = new ItemData(simple_hair);
    character.wear(hair.subType, hair.id, hair.amputationData);
    expect(Object.keys(character.clothes)).toHaveLength(1);
    expect(character.clothes[hair.subType]).toEqual(hair.id);
    expect(character).toMatchSnapshot();
  });

  test('CHARACTER: wear() - success: wearing a posed dress - assert limbs get amputated', () => {
    const character: Character = new Character();
    const dress: ItemData = new ItemData(posed_dress);
    character.wear(dress.subType, dress.id, dress.amputationData);
    validateAmputations(dress, character);
  });

  test('CHARACTER: wear() - success: prevent taking off hair if wearing the hair already', () => {
    const character: Character = new Character();
    const hair: ItemData = new ItemData(simple_hair);
    character.wear(hair.subType, hair.id, hair.amputationData);
    character.wear(hair.subType, hair.id, hair.amputationData);
    expect(character.clothes[hair.subType]).toBe(hair.id);
  });

  test('CHARACTER: wear() - success: replacing dress on character', () => {
    const character: Character = new Character();
    const originalBody: Body = new Set(character.body);

    // Wear a posed dress, validate that the limbs get amputated
    const dress1: ItemData = new ItemData(posed_dress);
    character.wear(dress1.subType, dress1.id, dress1.amputationData);
    expect(Object.keys(character.clothes)).toHaveLength(2);
    expect(character.clothes[dress1.subType]).toEqual(dress1.id);
    validateAmputations(dress1, character);
    expect(character).toMatchSnapshot();

    // Assert amputation data gets removed because we're replacing posed dress with simple dress
    const dress2: ItemData = new ItemData(simple_dress);
    character.wear(dress2.subType, dress2.id, dress2.amputationData);
    expect(Object.keys(character.clothes)).toHaveLength(2);
    expect(character.clothes[dress2.subType]).toEqual(dress2.id);

    // Assert body parts are made visible again when posed dress is removed
    expect(character.body).toEqual(originalBody);
    validateAmputations(dress2, character);
    expect(character).toMatchSnapshot();
  });

  test('CHARACTER: wear() - success: wearing multiple items that amputate the same body parts', () => {
    const character: Character = new Character();

    const dress: ItemData = new ItemData(posed_dress);
    character.wear(dress.subType, dress.id, dress.amputationData);
    validateAmputations(dress, character);
    expect(character).toMatchSnapshot();

    const coat: ItemData = new ItemData(posed_coat);
    character.wear(coat.subType, coat.id, coat.amputationData);
    validateAmputations(coat, character);
    expect(character).toMatchSnapshot();

    character.remove(SUBTYPES.DRESS as SubType);
    validateAmputations(dress, character);
    expect(character).toMatchSnapshot();
  });

  test('CHARACTER: remove() - success: wear and unwear the same item, should be back to where you started', () => {
    const character: Character = new Character();
    const copy: Character = new Character(character);
    
    const dress1: ItemData = new ItemData(posed_dress);
    character.wear(dress1.subType, dress1.id, dress1.amputationData);
    character.remove(SUBTYPES.DRESS as SubType);

    expect(character.clothes).toStrictEqual(copy.clothes);
    expect(character.amputations).toStrictEqual(copy.amputations);
    expect(character.body).toStrictEqual(copy.body);
  });

  test('CHARACTER: remove() - success: does nothing if removing nothing', () => {
    const character: Character = new Character();
    expect(character).toMatchSnapshot();
    character.remove(SUBTYPES.BOTTOM as SubType);
    expect(character).toMatchSnapshot();
  });
  
  test('Successfully creates a new Character out of existing Character', () => {
    const character: Character = new Character();    
    const dress1: ItemData = new ItemData(posed_dress);
    character.wear(dress1.subType, dress1.id, dress1.amputationData);

    const copy: Character = new Character(character);
    expect(character.clothes).toStrictEqual(copy.clothes);
    expect(character.amputations).toStrictEqual(copy.amputations);
    expect(character.body).toStrictEqual(copy.body);
  });
});

// describe('DataState', () => {
//   let store: Store<any>;
//   let apiMock: any;

//   beforeEach(() => {
//     store = createStoreWithMiddleware();
//     apiMock = jest.spyOn(api, 'fetchItemData');
//     apiMock.mockClear();
//   });

//   test('Action: ADD_ITEMS / Use-case: loadItem -- document does not exist', async () => {
//     apiMock.mockImplementation((itemId: number) => {}); // Document does not exist
//     await store.dispatch<any>(loadItem(22008));
//     const state: RootState = store.getState();

//     expect(apiMock).toBeCalledTimes(1);
//     expect(state.data.itemsData).toStrictEqual({});
//     expect(state.data.itemsData[22008]).toStrictEqual(undefined);
//   });
// });
