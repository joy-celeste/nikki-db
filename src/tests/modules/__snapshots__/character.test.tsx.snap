// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character CHARACTER: remove() - success: does nothing if removing nothing 1`] = `
Character {
  "amputations": Object {
    "10": Array [],
    "12": Array [],
    "5": Array [],
    "9": Array [],
  },
  "clothes": Object {
    "1": 10001,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    9,
    10,
    11,
    12,
  },
}
`;

exports[`Character CHARACTER: remove() - success: does nothing if removing nothing 2`] = `
Character {
  "amputations": Object {
    "10": Array [],
    "12": Array [],
    "5": Array [],
    "9": Array [],
  },
  "clothes": Object {
    "1": 10001,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    9,
    10,
    11,
    12,
  },
}
`;

exports[`Character CHARACTER: wear() - success case: dress on new character 1`] = `
Character {
  "amputations": Object {
    "10": Array [
      22008,
    ],
    "12": Array [
      22008,
    ],
    "5": Array [],
    "9": Array [
      22008,
    ],
  },
  "clothes": Object {
    "1": 10001,
    "2": 22008,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    11,
  },
}
`;

exports[`Character CHARACTER: wear() - success: replacing default hair with something else 1`] = `
Character {
  "amputations": Object {
    "10": Array [],
    "12": Array [],
    "5": Array [],
    "9": Array [],
  },
  "clothes": Object {
    "1": 10007,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    9,
    10,
    11,
    12,
  },
}
`;

exports[`Character CHARACTER: wear() - success: replacing dress on character 1`] = `
Character {
  "amputations": Object {
    "10": Array [
      22008,
    ],
    "12": Array [
      22008,
    ],
    "5": Array [],
    "9": Array [
      22008,
    ],
  },
  "clothes": Object {
    "1": 10001,
    "2": 22008,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    11,
  },
}
`;

exports[`Character CHARACTER: wear() - success: replacing dress on character 2`] = `
Character {
  "amputations": Object {
    "10": Array [],
    "12": Array [],
    "5": Array [],
    "9": Array [],
  },
  "clothes": Object {
    "1": 10001,
    "2": 21305,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    11,
    9,
    10,
    12,
  },
}
`;

exports[`Character CHARACTER: wear() - success: wearing multiple items that amputate the same body parts 1`] = `
Character {
  "amputations": Object {
    "10": Array [
      22008,
    ],
    "12": Array [
      22008,
    ],
    "5": Array [],
    "9": Array [
      22008,
    ],
  },
  "clothes": Object {
    "1": 10001,
    "2": 22008,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    11,
  },
}
`;

exports[`Character CHARACTER: wear() - success: wearing multiple items that amputate the same body parts 2`] = `
Character {
  "amputations": Object {
    "10": Array [
      22008,
    ],
    "12": Array [
      22008,
    ],
    "5": Array [],
    "9": Array [
      22008,
      30987,
    ],
  },
  "clothes": Object {
    "1": 10001,
    "2": 22008,
    "3": 30987,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    11,
  },
}
`;

exports[`Character CHARACTER: wear() - success: wearing multiple items that amputate the same body parts 3`] = `
Character {
  "amputations": Object {
    "10": Array [],
    "12": Array [],
    "5": Array [],
    "9": Array [
      30987,
    ],
  },
  "clothes": Object {
    "1": 10001,
    "3": 30987,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    11,
    10,
    12,
  },
}
`;

exports[`Character Initializes to default data if empty constructor 1`] = `
Character {
  "amputations": Object {
    "10": Array [],
    "12": Array [],
    "5": Array [],
    "9": Array [],
  },
  "clothes": Object {
    "1": 10001,
  },
  "visibleParts": Set {
    2,
    4,
    5,
    9,
    10,
    11,
    12,
  },
}
`;

exports[`CharacterState Action: CHARACTER_ADD_TO_HISTORY / Use-case: wearItem -- successfully add to history - wearing, replacing, unwearing items 1`] = `
Object {
  "character": Object {
    "history": Array [
      Character {
        "amputations": Object {
          "10": Array [],
          "12": Array [],
          "5": Array [],
          "9": Array [],
        },
        "clothes": Object {
          "1": 10001,
        },
        "visibleParts": Set {
          2,
          4,
          5,
          9,
          10,
          11,
          12,
        },
      },
      Character {
        "amputations": Object {
          "10": Array [],
          "12": Array [],
          "5": Array [],
          "9": Array [],
        },
        "clothes": Object {
          "1": 10007,
        },
        "visibleParts": Set {
          2,
          4,
          5,
          9,
          10,
          11,
          12,
        },
      },
      Character {
        "amputations": Object {
          "10": Array [
            22008,
          ],
          "12": Array [
            22008,
          ],
          "5": Array [],
          "9": Array [
            22008,
          ],
        },
        "clothes": Object {
          "1": 10007,
          "2": 22008,
        },
        "visibleParts": Set {
          2,
          4,
          5,
          11,
        },
      },
      Character {
        "amputations": Object {
          "10": Array [],
          "12": Array [],
          "5": Array [],
          "9": Array [],
        },
        "clothes": Object {
          "1": 10007,
          "2": 21305,
        },
        "visibleParts": Set {
          2,
          4,
          5,
          11,
          9,
          10,
          12,
        },
      },
      Character {
        "amputations": Object {
          "10": Array [],
          "12": Array [],
          "5": Array [],
          "9": Array [],
        },
        "clothes": Object {
          "1": 10007,
        },
        "visibleParts": Set {
          2,
          4,
          5,
          11,
          9,
          10,
          12,
        },
      },
      Character {
        "amputations": Object {
          "10": Array [],
          "12": Array [],
          "5": Array [],
          "9": Array [
            30987,
          ],
        },
        "clothes": Object {
          "1": 10007,
          "3": 30987,
        },
        "visibleParts": Set {
          2,
          4,
          5,
          11,
          10,
          12,
        },
      },
    ],
    "step": 5,
  },
  "data": Object {
    "itemsData": Object {
      "10007": ItemData {
        "amputationData": null,
        "depthType": 1,
        "depths": Object {
          "1": 33000,
          "2": 3500,
          "7": 38000,
          "8": 37200,
          "9": 3510,
        },
        "description": "A gentle, thoughtful nobleman is not supposed to be the silent background forever.",
        "id": 10007,
        "name": "Elegant Nobleman",
        "position": Array [
          PositionData {
            "depth": 1,
            "scale": null,
            "x": -21.5,
            "y": 426,
            "z": 33000,
          },
        ],
        "subType": 1,
      },
      "21305": ItemData {
        "amputationData": null,
        "depthType": 3,
        "depths": Object {
          "1": 24000,
          "2": 17000,
          "3": 6000,
          "9": 10220,
        },
        "description": "In the pseudonym of Zisang, Bai Jinjin handed in a special white and red Apple style dress this time.",
        "id": 21305,
        "name": "Hiding in the WorldÂ·Red",
        "position": Array [
          PositionData {
            "depth": 1,
            "scale": 1,
            "x": -7.5,
            "y": 120,
            "z": 24000,
          },
        ],
        "subType": 2,
      },
      "22008": ItemData {
        "amputationData": Object {
          "10": true,
          "12": true,
          "5": false,
          "9": true,
        },
        "depthType": 41,
        "depths": Object {
          "1": 36600,
          "12": 10200,
          "13": 10170,
          "14": 9900,
          "2": 24000,
          "21": 24000,
          "22": 24000,
          "3": 18000,
          "4": 6000,
          "5": 850,
          "8": 27150,
          "9": 10220,
        },
        "description": "The sakura festival has come again. In the flower rain, she holds an umbrella, waiting for the one to come.",
        "id": 22008,
        "name": "Sakura Dream",
        "position": Array [
          PositionData {
            "depth": 2,
            "scale": 1,
            "x": -32.5,
            "y": -164.5,
            "z": 24000,
          },
          PositionData {
            "depth": 901,
            "scale": 1,
            "x": -59.5,
            "y": 328,
            "z": 10215,
          },
        ],
        "subType": 2,
      },
      "30987": ItemData {
        "amputationData": Object {
          "10": false,
          "12": false,
          "5": false,
          "9": true,
        },
        "depthType": 48,
        "depths": Object {
          "1": 36700,
          "12": 10190,
          "13": 10160,
          "14": 9880,
          "2": 27000,
          "3": 19000,
          "4": 5000,
          "5": 840,
        },
        "description": "He stands against the light and looks back at her. She saves the throb deeply into her heart.",
        "id": 30987,
        "name": "Rose Heart",
        "position": Array [
          PositionData {
            "depth": 1,
            "scale": 1,
            "x": -27,
            "y": 200.5,
            "z": 36700,
          },
          PositionData {
            "depth": 2,
            "scale": 1,
            "x": -202,
            "y": -130,
            "z": 27000,
          },
          PositionData {
            "depth": 3,
            "scale": 1,
            "x": 101,
            "y": -119,
            "z": 19000,
          },
          PositionData {
            "depth": 4,
            "scale": 1,
            "x": -81,
            "y": -153,
            "z": 5000,
          },
          PositionData {
            "depth": 13,
            "scale": 1,
            "x": -88.5,
            "y": 245,
            "z": 10160,
          },
          PositionData {
            "depth": 14,
            "scale": 1,
            "x": 42.5,
            "y": 234.5,
            "z": 9880,
          },
          PositionData {
            "depth": 101,
            "scale": 1,
            "x": -30.5,
            "y": 242.5,
            "z": 36695,
          },
          PositionData {
            "depth": 201,
            "scale": 1,
            "x": -47.5,
            "y": 262,
            "z": 26995,
          },
          PositionData {
            "depth": 301,
            "scale": 1,
            "x": 124.5,
            "y": 139,
            "z": 18995,
          },
        ],
        "subType": 3,
      },
    },
    "loading": false,
  },
}
`;

exports[`CharacterState Assert initial state uses default values 1`] = `
Object {
  "history": Array [
    Character {
      "amputations": Object {
        "10": Array [],
        "12": Array [],
        "5": Array [],
        "9": Array [],
      },
      "clothes": Object {
        "1": 10001,
      },
      "visibleParts": Set {
        2,
        4,
        5,
        9,
        10,
        11,
        12,
      },
    },
  ],
  "step": 0,
}
`;
